diff --git a/thumbs.c b/thumbs.c
index 209c94b..e2f27f4 100644
--- a/thumbs.c
+++ b/thumbs.c
@@ -38,6 +38,19 @@ void exif_auto_orientate(const fileinfo_t*);
 
 static char *cache_dir;
 
+const char* get_file_ext(const char path[])
+{
+	const char *type = strrchr(path, '.');
+	if (type != NULL && type != path)
+		return type + 1;
+	return "";
+}
+
+const char* jpg_or_png(const char type[])
+{
+	return (strcmp(type, "png") && strcmp(type, "gif")) ? "jpg" : "png";
+}
+
 char* tns_cache_filepath(const char *filepath)
 {
 	size_t len;
@@ -50,7 +63,7 @@ char* tns_cache_filepath(const char *filepath)
 		/* don't cache images inside the cache directory! */
 		len = strlen(cache_dir) + strlen(filepath) + 6;
 		cfile = (char*) s_malloc(len);
-		snprintf(cfile, len, "%s/%s.jpg", cache_dir, filepath + 1);
+		snprintf(cfile, len, "%s/%s.%s", cache_dir, filepath + 1, jpg_or_png(get_file_ext(filepath)));
 	}
 	return cfile;
 }
@@ -101,7 +114,7 @@ void tns_cache_write(Imlib_Image im, const char *filepath, bool force)
 			}
 			if (err == 0) {
 				imlib_context_set_image(im);
-				imlib_image_set_format("jpg");
+				imlib_image_set_format(jpg_or_png(get_file_ext(filepath)));
 				imlib_image_attach_data_value("quality", NULL, 90, NULL);
 				imlib_save_image_with_error_return(cfile, &err);
 			}
