diff --git a/thumbs.c b/thumbs.c
index 209c94b..21a4edb 100644
--- a/thumbs.c
+++ b/thumbs.c
@@ -217,20 +217,19 @@ void tns_free(tns_t *tns)
 Imlib_Image tns_scale_down(Imlib_Image im, int dim)
 {
 	int w, h;
-	float z, zw, zh;
 
 	imlib_context_set_image(im);
 	w = imlib_image_get_width();
 	h = imlib_image_get_height();
-	zw = (float) dim / (float) w;
-	zh = (float) dim / (float) h;
-	z = MIN(zw, zh);
-	z = MIN(z, 1.0);
 
-	if (z < 1.0) {
+	int x = (w < h) ? 0 : (w - h) / 2;
+	int y = (w > h) ? 0 : (h - w) / 2;
+
+	int s = (w < h) ? w : h;
+
+	if (dim < w || dim < h) {
 		imlib_context_set_anti_alias(1);
-		im = imlib_create_cropped_scaled_image(0, 0, w, h,
-		                                       MAX(z * w, 1), MAX(z * h, 1));
+		im = imlib_create_cropped_scaled_image(x, y, s, s, dim, dim);
 		if (im == NULL)
 			die("could not allocate memory");
 		imlib_free_image_and_decache();
